### ISSUE TITLE
BLT and loading customized database on ACE

### ISSUE THREAD

**Issue by rgristroph** [acquia/blt#1685]
My system information:
* Operating system type: Ubuntu 12.04.5 LTS
* Operating system version: Ubuntu 12.04.5 LTS
* BLT version: 8.7.3

Output of `blt doctor`:
```
+----------------------+----------------------------------------------------------------------------------+
| Property             | Value                                                                            |
+----------------------+----------------------------------------------------------------------------------+
| %paths               |                                                                                  |
|  - %root             | /mnt/gfs/XXXXXXX/livedev/docroot                                          |
|  - %site             | sites/default                                                                    |
|  - %modules          | modules                                                                          |
|  - %themes           | themes                                                                           |
|  - %config-sync      | /mnt/gfs/XXXXXXX/livedev/docroot/sites/default/files/config_566456107dd0b |
|                      | ad1a00d4ab6b1d5ca7b21d290c9/sync                                                 |
|  - %config-vcs       | /mnt/gfs/XXXXXXXX/livedev/docroot/../config/default                        |
|  - %files            | sites/default/files                                                              |
|  - %private          | /mnt/files/XXXXXXXX/files-private                                         |
|  - %temp             | /mnt/tmp/XXXXXXX                                                            |
| admin-theme          | seven                                                                            |
| blt-version          | 8.7.3                                                                            |
| bootstrap            | Successful                                                                       |
| config-sync          | /mnt/gfs/XXXXXXXXXX/livedev/docroot/sites/default/files/config_566456107dd0b |
|                      | ad1a00d4ab6b1d5ca7b21d290c9/sync                                                 |
| config-vcs           | /mnt/gfs/XXXXXXXXXXXXXX/livedev/docroot/../config/default                        |
| db-driver            | mysql                                                                            |
| db-hostname          | dbmaster-NNNNN                                                                  |
| db-name              | XXXXXXXXXXXXX                                                                 |
| db-password          | XXXXXXXXXXXX                                                                |
| db-port              | 3306                                                                             |
| db-status            | Connected                                                                        |
| db-username          | XXXXXXXXX                                                                           |
| drupal-settings-file | sites/default/settings.php                                                       |
| drupal-version       | 8.3.1                                                                            |
| drush-alias-files    | /mnt/drush/XXXXXX/XXXXXXX.aliases.drushrc.php                                     |
|                      | /mnt/gfs/XXXXXXXXXXX/livedev/docroot/../drush/site-aliases/aliases.drushrc.ph |
|                      | p                                                                                |
|                      | /mnt/gfs/iXXXXXXXXXXX/livedev/docroot/../drush/site-aliases/XXXX.aliases.drushr |
|                      | c.php                                                                            |
| drush-conf           | /mnt/gfs/XXXXXXXXX/livedev/drush/drushrc.php                                |
|                      | /etc/drush/drushrc.php                                                           |
| drush-script         | /mnt/gfs/xXXXXXXXX/livedev/vendor/drush/drush/drush.php                     |
| drush-temp           | /mnt/tmp/XXXXXXXXX                                                         |
| drush-version        | 8.1.11                                                                           |
| files                | sites/default/files                                                              |
| install-profile      | standard                                                                         |
| modules              | modules                                                                          |
| php-bin              | /usr/local/php5.6/bin/php                                                        |
| php-conf             | /usr/local/php5.6/etc/cli/php.ini                                                |
| php-os               | Linux                                                                            |
| private              | /mnt/files/XXXXXXX/files-private                                         |
| root                 | /mnt/gfs/XXXXXXX/livedev/docroot                                          |
| site                 | sites/default                                                                    |
| temp                 | /mnt/tmp/XXXXXXXt                                                            |
| theme                | XXXXXXXX                                                                        |
| themes               | themes                                                                           |
| uri                  | http://default                                                                   |
| user                 |                                                                                  |
+----------------------+----------------------------------------------------------------------------------+
+--------------------------+----------------------------------------------------------------------------------+
| Check                    | Outcome                                                                          |
+--------------------------+----------------------------------------------------------------------------------+
| checkDocrootExists       | Found docroot.                                                                   |
| checkCoreExists          | Drupal core exists                                                               |
| checkSettingsFile        | BLT settings are included in settings file.                                      |
| checkLocalSettingsFile   | Could not find local settings file.                                              |
|                          | Your local settings file should exist at                                         |
|                          | /mnt/gfs/XXXXXXXXXX/livedev/docroot/sites/default/settings/local.settings.ph |
|                          | p.                                                                               |
| checkLocalDrushFile      | Local drushrc file does not exist.                                               |
|                          | Create /mnt/gfs/XXXXXXX/livedev/docroot/sites/default/local.drushrc.php.  |
|                          | Run `blt setup:drush:settings` to generate it automatically, or run `blt setup`  |
|                          | to run the entire setup process.                                                 |
| checkFileSystem:%files   | Public files directory is writable.                                              |
| checkFileSystem:%private | Private files directory is writable.                                             |
| checkFileSystem:%temp    | Temporary files directory is writable.                                           |
| checkDbConnection        | Connected to database.                                                           |
| checkDrupalBootstrapped  | Bootstrapped Drupal via drush.                                                   |
| checkDrupalInstalled     | Drupal is installed.                                                             |
| checkCaching:page        | Drupal cache is disabled.                                                        |
| checkCaching:css         | CSS preprocessing is disabled.                                                   |
| checkCaching:js          | JS preprocessing is disabled.                                                    |
| checkNvmExists           | NVM does not exist.                                                              |
|                          |                                                                                  |
|                          | It is recommended that you use NVM to manage multiple versions of NodeJS on one  |
|                          | machine.                                                                         |
|                          | Instructions for installing NVM can be found at:                                 |
|                          |   https://github.com/creationix/nvm#installation                                 |
| checkDevDesktop          | DevDesktop usage is enabled, but $DEVDESKTOP_DRUPAL_SETTINGS_DIR is not set in   |
|                          | your environmental variables.                                                    |
|                          |                                                                                  |
|                          | Add `export                                                                      |
|                          | DEVDESKTOP_DRUPAL_SETTINGS_DIR="$HOME/.acquia/DevDesktop/DrupalSettings"` to     |
|                          | ~/.bash_profile or equivalent for your system.`                                  |
| checkDevDesktop          | variables_order allows environment variables in php.ini.                         |
| checkComposer:require    | acquia/blt is in composer.json's require object.                                 |
| checkComposer:plugins    | hirak/prestissimo plugin for composer is not installed.                          |
|                          |   Run `composer global require hirak/prestissimo:^0.3` to install it.            |
|                          |   This will improve composer install/update performance by parallelizing the     |
|                          | download of dependency information.                                              |
| checkBehat:exists        | tests/behat/local.yml is missing!                                                |
|                          |   Run `blt setup:behat` to generate it from example.local.yml.                   |
| checkProjectYml:keys     | project.yml has no deprecated keys.                                              |
| checkAcsfConfig          | BLT settings are not included in your pre-settings-php include.                  |
|                          |   Add a require statement for "/../vendor/acquia/blt/settings/blt.settings.php"  |
|                          | to                                                                               |
|                          | /mnt/gfs/XXXXXXXX/livedev/docroot/../factory-hooks/pre-settings-php/includ |
|                          | es.php                                                                           |
| checkPhpDateTimezone     | PHP setting for date.timezone is correctly set                                   |
+--------------------------+----------------------------------------------------------------------------------+
```

We are on Acquia hosting, and recently switched from BLT 8.3.3 to 8.7.3.  We have a customization to the $databases array, to force the use of a particular database on a particular environment, that previously was in docroot/site/default/settings/blt.settings.php and move to docroot/site/default/settings/includes.settings.php with the move to 8.7.3.

The problem is we lost the ability to override the $databases array with 8.7.3.  What is surprising, is that even if I "print_r($databases);" at the end of the settings.php file, I see our override correctly in there . . . but drush status shows that AFTER that, the default DB config is somehow restored.

I would think this was a special configuration of Acquia hosting, except that it works as expected with the older BLT.

**Comment by grasmash** [comment]
A few questions:

> the default DB config is somehow restored

* Can you reproduce this on your local machine?

> We have a customization to the $databases array

Can you share a snippet?

>  but drush status shows that AFTER that,

Are you executing `drush status` in the docroot?

**Comment by grasmash** [comment]
Just double checking, you wrote:

> docroot/site/default/settings/includes.settings.php

Did you mean `/sites/default/settings/includes.settings.php` ?

**Comment by rgristroph** [comment]
Reproducing on a local machine -- let me try that.

Snippet customizing the $databases array --

```
if ($is_ah_env && file_exists('/var/www/site-php')) {
  // Allow the xxxRENDER environment to use the xxxxEDIT database.
  if ($_ENV['AH_SITE_ENVIRONMENT'] == 'xxxrender' && $_ENV['AH_SITE_GROUP'] == 'ZZZZ') {
    $edit_db_include = '/home/XXXXXX/var-www-site-php-XXXXXXXedit/D8-XXedit-XXXXX-settings.inc';
    if (file_exists($edit_db_include)) {
      include($edit_db_include);
    }
  }
  elseif ($_ENV['AH_SITE_ENVIRONMENT'] == 'XXXXrendermr' && $_ENV['AH_SITE_GROUP'] == 'XXXXXXX') {
    $edit_db_include = '/var/www/site-php/XXXXXX.XXXXeditmr/D8-XXXeditmr-XXXXXX-settings.inc';
    if (file_exists($edit_db_include)) {
      $render_trusted_host_patterns = $settings['trusted_host_patterns'];
      include($edit_db_include);
      $settings['trusted_host_patterns'] = $render_trusted_host_patterns;
    }
  }
  else {
    require_once "/var/www/site-php/{$_ENV['AH_SITE_GROUP']}/{$_ENV['AH_SITE_GROUP']}-settings.inc";
  }
```

Running the drush status in the docroot -- I am, it seems to make no difference.

I did mean /sites/default/settings/includes.settings.php .

**Comment by grasmash** [comment]
Seems like this had to do with Acquia Hosting settings rather than BLT. @rgristroph is using:

```
unset($conf['acquia_hosting_site_info']);
unset($databases);
$conf['acquia_hosting_settings_autoconnect'] = FALSE;
if ($db_connection = Database::getConnection() ) {
  Database::removeConnection($db_connection->getKey());
}
```

@rgristroph Please update with any other detail you have. Closing.

**Comment by rgristroph** [comment]
Just an update on this, which turned out not to be a problem with BLT but the solution might help someone.

In the older BLT, the blt.settings.php file was in the docroot, and we had modified the part where it did ```require_once "/var/www/site-php/{$_ENV['AH_SITE_GROUP']}/{$_ENV['AH_SITE_GROUP']}-settings.inc";``` to load a different file.  The key thing is the file was only required once.  In the newer BLT, the blt.settings.php file is in /vendor/acquia/blt/settings/ , and we had moved our custom logic to the sites/default/settings/includes.settings.php file, which means that a database credentials file was sourced twice -- once the /vendor version sourced the one you are supposed to use on the hosting platform, and then again a different one was sourced.

After a bit of testing I figured out the main problem was that when one of those db credentials files is sourced, it initiates a DB connection to test the credentials.  If that's still there it won't switch to the new one when the second db credentials file is loaded.  Thus, the "unset" and "acquia_hosting_settings_autoconnect" lines in the above snippet aren't actually necessary -- the key is, right before the 2d db credentials file is sourced, that we want to override the default, have these lines:

```
if ($db_connection = Database::getConnection() ) {
  Database::removeConnection($db_connection->getKey());
}
```
Also at the top of includes.settings.php you will need

```
use Drupal\Core\Database\Database;
```
